lightning/src/ln/channelmanager.rs

```diff
@@ -355,15 +359,32 @@ struct ReceiveError {
 pub enum FailureCode {
 	/// We had a temporary error processing the payment. Useful if no other error codes fit
 	/// and you want to indicate that the payer may want to retry.
-	TemporaryNodeFailure             = 0x2000 | 2,
+	TemporaryNodeFailure,
 	/// We have a required feature which was not in this onion. For example, you may require
 	/// some additional metadata that was not provided with this payment.
-	RequiredNodeFeatureMissing       = 0x4000 | 0x2000 | 3,
+	RequiredNodeFeatureMissing,
 	/// You may wish to use this when a `payment_preimage` is unknown, or the CLTV expiry of
 	/// the HTLC is too close to the current block height for safe handling.
 	/// Using this failure code in [`ChannelManager::fail_htlc_backwards_with_reason`] is
 	/// equivalent to calling [`ChannelManager::fail_htlc_backwards`].
-	IncorrectOrUnknownPaymentDetails = 0x4000 | 15,
+	IncorrectOrUnknownPaymentDetails,
+	/// We failed to process the payload after the onion was decrypted. You may wish to
+	/// use this when receiving custom HTLC TLVs with even type numbers that you don't recognize.
```

Comment:
(a) I think deanonymization attack based on onion failure message is in LN security model, see the now-deprecated `incorrect_payment_amount` and `final_expiry_too_soon and b) You introduce a probing vector between versions of LDK, the one giving you back an unknown TLV record and the one giving you a valid result for the same custom TLV record (independently of this being a success or error).

In a world with blinded paths, a payer might never be aware of the node_pubkey of the payee, and such malicious onion payload is a valid deanonymization vector. I agree with you in a world pre-blinding paths, we would care less about fingerprinting.

Reply:
> That's a different attack - that's revealing if an HTLC is yours vs if its not.

See the other point if you introduce an overlay protocol like cancellable payment, what content you return as an onion message might leak internal state machine, and therefore if it’s revealing if a HTLC is yours.

On the second attack, software user agent and version are considered in browser deanomymization literature (cf. https://leakuidatorplusteam.github.io/preprint.pdf) and it’s a strong technique to break the one whole anonymous set in segment. As a reminder as a successful de-anonymization attack is just multiple cumulating multiple leaks and correlating them until you can probe the result.

Reply:
> (https://github.com/lightningdevkit/rust-lightning/pull/2308#discussion_r1278165858) I agree with you in a world pre-blinding paths, we would care less about fingerprinting.

> (https://github.com/lightningdevkit/rust-lightning/pull/2308#issuecomment-1659476435) yes, though i’m not sure no returning error is workable on the long-term due to onion bandwidth cost and payment reliability (at least end-to-end), though here more conversation at the spec-level.

Since we return no error for blinded paths (which may be up to further discussion upstream), and pre-blinded paths fingerprinting is outside our security model, is there any action here?

Reply:
> Since we return no error for blinded paths (which may be up to further discussion upstream), and pre-blinded paths fingerprinting is outside our security model, is there any action here?

Looking on the code, it sounds we always return a silent `invalid_onion_payload` as suggested i.e `InvalidOnionPayload(None)` here: https://github.com/lightningdevkit/rust-lightning/pull/2308/files#diff-645d12c7269d09caab57d2e0c1fd34f672ee259994c86ad4bad7f4da9183671cR4812

So I think it’s good in case of future errors return even for blinded paths in the future we’re good, no need to change the code.

Looking on the documentation of `FailureCode::InvalidOnionPayload`, “if available, the tuple data should include the type number and byte offset in the decrypted byte stream where the failure occurred”. This could be change to say we never return the custom TLV to avoid leaking user agent / version in case of future onion changes.

Reply:
Just changed to say "may" instead of "should". I'm going to hold off on mentioning leaking user agent/version unless another reviewer chimes in as it still seems to be outside our security model from what I understand

Reply:
Well “may” is reasonable and PR should be numbered in release notes, so user can always go back to this conversation X years from now, assuming GH as a platform is still available.